File: MVC y DAO/src/main/java/co/edu/unisalle/crud/dao/EstudianteDAOJson.java
package co.edu.unisalle.crud.dao;

import co.edu.unisalle.crud.model.Estudiante;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

import java.io.File;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Implementación del DAO que persiste estudiantes en un archivo JSON.
 *
 * SRP: única responsabilidad — leer y escribir estudiantes en JSON.
 * LSP: respeta el contrato declarado por {@link IEstudianteDAO}; cualquier
 * implementación puede sustituirla sin romper a los clientes.
 */
public class EstudianteDAOJson implements IEstudianteDAO {

    private final File archivo;
    private final ObjectMapper mapper;

    public EstudianteDAOJson(File archivo) {
        this.archivo = archivo;
        this.mapper = new ObjectMapper();
        this.mapper.enable(SerializationFeature.INDENT_OUTPUT);
        inicializarArchivoSiNoExiste();
    }

    private void inicializarArchivoSiNoExiste() {
        if (!archivo.exists()) {
            File padre = archivo.getParentFile();
            if (padre != null && !padre.exists()) {
                padre.mkdirs();
            }
            escribirTodos(new ArrayList<>());
        }
    }

    @Override
    public Estudiante crear(Estudiante estudiante) {
        if (estudiante == null) {
            throw new IllegalArgumentException("El estudiante no puede ser nulo.");
        }
        List<Estudiante> estudiantes = listar();
        boolean existe = estudiantes.stream().anyMatch(e -> e.getId() == estudiante.getId());
        if (existe) {
            throw new IllegalArgumentException(
                    "Ya existe un estudiante con id " + estudiante.getId());
        }
        estudiantes.add(estudiante);
        escribirTodos(estudiantes);
        return estudiante;
    }

    @Override
    public List<Estudiante> listar() {
        try {
            if (archivo.length() == 0) {
                return new ArrayList<>();
            }
            return mapper.readValue(archivo, new TypeReference<List<Estudiante>>() {});
        } catch (IOException e) {
            throw new UncheckedIOException("No fue posible leer el archivo JSON.", e);
        }
    }

    @Override
    public Optional<Estudiante> buscarPorId(int id) {
        return listar().stream()
                .filter(e -> e.getId() == id)
                .findFirst();
    }

    @Override
    public Estudiante actualizar(Estudiante estudiante) {
        if (estudiante == null) {
            throw new IllegalArgumentException("El estudiante no puede ser nulo.");
        }
        List<Estudiante> estudiantes = listar();
        for (int i = 0; i < estudiantes.size(); i++) {
            if (estudiantes.get(i).getId() == estudiante.getId()) {
                estudiantes.set(i, estudiante);
                escribirTodos(estudiantes);
                return estudiante;
            }
        }
        throw new IllegalArgumentException(
                "No existe un estudiante con id " + estudiante.getId());
    }

    @Override
    public boolean eliminar(int id) {
        List<Estudiante> estudiantes = listar();
        boolean eliminado = estudiantes.removeIf(e -> e.getId() == id);
        if (eliminado) {
            escribirTodos(estudiantes);
        }
        return eliminado;
    }

    private void escribirTodos(List<Estudiante> estudiantes) {
        try {
            mapper.writeValue(archivo, estudiantes);
        } catch (IOException e) {
            throw new UncheckedIOException("No fue posible escribir el archivo JSON.", e);
        }
    }
}

File: MVC y DAO/CrudMvcDao/Dao/EstudianteDaoJson.cs
using System.Text.Json;
using CrudMvcDao.Model;

namespace CrudMvcDao.Dao;

/// <summary>
/// Implementación del DAO que persiste estudiantes en un archivo JSON.
///
/// SRP: única responsabilidad — leer y escribir estudiantes en JSON.
/// LSP: respeta el contrato declarado por <see cref="IEstudianteDao"/>;
/// cualquier implementación puede sustituirla sin romper a los clientes.
/// </summary>
public class EstudianteDaoJson : IEstudianteDao
{
    private readonly string _archivoPath;
    private readonly JsonSerializerOptions _options;

    public EstudianteDaoJson(string archivoPath)
    {
        _archivoPath = archivoPath;
        _options = new JsonSerializerOptions { WriteIndented = true };
        InicializarArchivoSiNoExiste();
    }

    private void InicializarArchivoSiNoExiste()
    {
        if (!File.Exists(_archivoPath))
        {
            var dir = Path.GetDirectoryName(_archivoPath);
            if (!string.IsNullOrEmpty(dir) && !Directory.Exists(dir))
                Directory.CreateDirectory(dir);
            EscribirTodos(new List<Estudiante>());
        }
    }

    public Estudiante Crear(Estudiante estudiante)
    {
        ArgumentNullException.ThrowIfNull(estudiante);
        var estudiantes = LeerTodos();
        if (estudiantes.Any(e => e.Id == estudiante.Id))
            throw new ArgumentException($"Ya existe un estudiante con id {estudiante.Id}");
        estudiantes.Add(estudiante);
        EscribirTodos(estudiantes);
        return estudiante;
    }

    public IReadOnlyList<Estudiante> Listar() => LeerTodos();

    public Estudiante? BuscarPorId(int id) =>
        LeerTodos().FirstOrDefault(e => e.Id == id);

    public Estudiante Actualizar(Estudiante estudiante)
    {
        ArgumentNullException.ThrowIfNull(estudiante);
        var estudiantes = LeerTodos();
        var idx = estudiantes.FindIndex(e => e.Id == estudiante.Id);
        if (idx < 0)
            throw new ArgumentException($"No existe un estudiante con id {estudiante.Id}");
        estudiantes[idx] = estudiante;
        EscribirTodos(estudiantes);
        return estudiante;
    }

    public bool Eliminar(int id)
    {
        var estudiantes = LeerTodos();
        var removidos = estudiantes.RemoveAll(e => e.Id == id);
        if (removidos > 0)
            EscribirTodos(estudiantes);
        return removidos > 0;
    }

    private List<Estudiante> LeerTodos()
    {
        if (!File.Exists(_archivoPath) || new FileInfo(_archivoPath).Length == 0)
            return new List<Estudiante>();
        var json = File.ReadAllText(_archivoPath);
        if (string.IsNullOrWhiteSpace(json))
            return new List<Estudiante>();
        return JsonSerializer.Deserialize<List<Estudiante>>(json, _options)
               ?? new List<Estudiante>();
    }

    private void EscribirTodos(List<Estudiante> estudiantes)
    {
        var json = JsonSerializer.Serialize(estudiantes, _options);
        File.WriteAllText(_archivoPath, json);
    }
}
